### PR TITLE
lintcheck: clippy fixes, test on ci

### PR DESCRIPTION
**Pull request by matthiaskrgr**
fixes clippy warnings in lintcheck and adds a small test (run on 3 small crates) to gha ci

changelog: none

### CODE REVIEW

**Comment by flip1995** [comment]
The `Clippy Dev` workflow hasn't installed rustc-dev-preview.

**Comment by matthiaskrgr** [comment]
Mmh... I removed the step that removes the replaces the `rust-toolchain` file with the `toolchain` action because I actually need to build and run clippy during the lintcheck test and it needs a matching (none-"nightly") version.

I don't think I can make the toolchain action use the pinned nightly somehow for installing rustfmt...?

**Comment by matthiaskrgr** [comment]
Ok, I think this works now:

We first remove the `rust-toolchain` file and grab an independent `rustfmt` version to check the formatting.
Then we restore the toolchain file via git and build clippy with the pinned nightly version.

**Comment by matthiaskrgr** [comment]
@bors try

just to be sure...

File: clippy_dev/src/lintcheck.rs [hunk 1]
```diff
@@ -328,9 +330,8 @@ impl LintcheckConfig {
         // by default use a single thread
         let max_jobs = match clap_config.value_of("threads") {
             Some(threads) => {
-                let threads: usize = threads
-                    .parse()
-                    .expect(&format!("Failed to parse '{}' to a digit", threads));
+                let err_msg = format!("Failed to parse '{}' to a digit", threads);
+                let threads: usize = threads.parse().expect(&err_msg);
```
Comment (camsteffen) [inline clippy_dev/src/lintcheck.rs:334]: Well...I think you're supposed to `unwrap_or_else` `panic!`

**Comment by flip1995** [comment]
WDYT about making a new workflow to check lintcheck and only run it, if `lintcheck.rs` is changed?

**Comment by matthiaskrgr** [comment]
if we only check lintcheck if `lintcheck.rs` is touched, we won't be able to find out when lintcheck breaks as it did when we had the `[workspace]`-related changes, which was my main motivation behind the test!

**Comment by flip1995** [comment]
Ah right, good point. CI changes LGTM then. Changes to lintcheck also seem to resolve almost all Clippy complaints. (I think the pedantic lints would still trigger on it though)

**Comment by matthiaskrgr** [comment]
Didn't see any pedantic warnings with `cargo run --bin cargo-clippy --manifest-path=rust-clippy/Cargo.toml  -- -Wclippy::pedantic  --features lintcheck`

File: .github/workflows/clippy_dev.yml [hunk 2]
```diff
@@ -21,35 +21,49 @@ jobs:
...
+          profile: minimal
+          components: rustfmt
+          default: true
+
+      - name: Test fmt
+        run: cargo dev fmt --check
+
+      # Restore the rust-toolchain-file
+      - name: Restore rust-toolchain-file
+        run: git checkout rust-toolchain
```
Comment (flip1995) [inline .github/workflows/clippy_dev.yml:46]: I would move this right before  the `lintcheck` test. That way clippy_dev has to be only rebuild once.

Currently it is built three times:

1. For rustfmt with +nightly
2. For the other tests with +<rust-toolchain-nightly>
3. For lintcheck --feature lintcheck

File: .github/workflows/clippy_dev.yml [hunk 3]
```diff
@@ -21,35 +21,49 @@ jobs:
...
+        uses: actions-rs/toolchain@v1.0.6
+        with:
+          toolchain: nightly
+          target: x86_64-unknown-linux-gnu
+          profile: minimal
+          components: rustfmt
+          default: true
+
+      - name: Test fmt
+        run: cargo dev fmt --check
```
Comment (flip1995) [inline .github/workflows/clippy_dev.yml:42]: And this should stay where it currently is, so it can profit from the `Build` job.

**Comment by flip1995** [comment]
I just wonder if the lintcheck test should be moved to the Clippy workflow to not have to build clippy again.

**Comment by flip1995** [comment]
@bors try

(hijacking this PR to fix up the Clippy CI.)

**Comment by flip1995** [comment]
Hmm.. I really don't like checking the lintcheck tool in CI. It will be compile-checked by dogfood, once this PR is done and I can continue to work on the dogfood fix. I don't think we should put 3-5min CI time on any workflow to run lintcheck tests though. Thoughts? @matthiaskrgr

**Comment by flip1995** [comment]
Also please review this commit: e465a8b 

IIUC the `--fix` test only checks that the `--fix` command is working? In that case it shouldn't matter on which crate we run the test on.

**Comment by matthiaskrgr** [comment]
>IIUC the --fix test only checks that the --fix command is working? In that case it shouldn't matter on which crate we run the test on

Yes, I think the `--fix` test was only added after https://github.com/rust-lang/rust-clippy/issues/6487 to make sure it does not break again. It probably does not really matter what crate is checked although it might be nice to have a 0-dep crate with an actually lint warning to fix inside :)

Building clippy-dev with the lintcheck feature probably takes around 1:30 minutes

Running the lintcheck test itself should take less than 30 seconds, with a prebuilt clippy binary
So yes, we should probably move it over to the config that already build clippy.

**Comment by flip1995** [comment]
The lintcheck job took 2m52s, which is approximately the same amount of time, that I got yesterday while testing. I'm still not convinced that that is worth it, since it is compile-checked by dogfood anyway.

> reformat clippy.ymk and clippy_bors.yml 458cf05 

The previous formatting was done according to the examples in https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions. I don't think it makes a difference, so I think reformatting to a yaml style that feels more natural (IMO) should be fine.

**Comment by matthiaskrgr** [comment]
I've split out the clippy lint fixes for lintcheck into https://github.com/rust-lang/rust-clippy/pull/6839, I think there was something waiting on that?

I guess I'll then rework this branch to not test lintcheck on ci, sigh :/

**Comment by flip1995** [comment]
> I think there was something waiting on that?

Yes my dogfood fix, thanks for splitting it out!

> I guess I'll then rework this branch to not test lintcheck on ci,  sigh :/

Sorry. I don't think 3min CI time is worth for testing lintcheck, since it is a rarely used dev-tool, I don't expect to break often, once it gets a bit more "stable". If I'm proven wrong in this regard, I'm of course happy to include it.

**Pull request closed without merging**